### PR TITLE
Editor update: Map import, merge. Export .bsp models. Export .obj.

### PR DESCRIPTION
**Pull request by 2020karaulov2020**
Editor update: Map import, merge. Export .bsp models. Export .obj.

Fixed setting lost (when force exit while settings is loading).
Fixed setting window crashes.

### CODE REVIEW

**Comment by 2020karaulov2020** [comment]
need add something more ... )

**Comment by 2020karaulov2020** [comment]
Editor update. Map import. Map merge. Export .obj.

**Comment by 2020karaulov2020** [comment]
@wootguy you can test it ?

Now editor can be run with command 'editor' or 'empty'. 
# Fix crashes with empty bsp
* Import maps (Import->Map file)
* Merge maps in one (Widgets->Map merge)
* Export to obj (now only geometry)

**Comment by 2020karaulov2020** [comment]
@wootguy 

[](http://www.youtube.com/watch?v=phs_9ScFSf4 "BSPGUY EDITOR MERGE FEATURE")

File: src/util/mat4x4.h [hunk 1]
```diff
@@ -28,20 +28,20 @@ struct mat4x4
...
 
-	float& operator ()(size_t idx)
-	{
-		return m[idx];
-	}
-
-	float operator ()(size_t idx) const
-	{
-		return m[idx];
+	float& operator ()(size_t idx)
```
Comment (2020karaulov2020) [inline src/util/mat4x4.h:31]: Fix windows + linux  newline ('\r\n and \n mixed)

File: src/util/util.cpp [hunk 2]
```diff
@@ -694,6 +694,28 @@ bool pointInsidePolygon(vector<vec2>& poly, vec2 p) {
 	return inside;
 }
 
+void tga_write(const char* filename, uint32_t width, uint32_t height, uint8_t* data, uint8_t dataChannels, uint8_t fileChannels)
```
Comment (2020karaulov2020) [inline src/util/util.cpp:697]: Untested (for export obj with textures, later)

**Comment by 2020karaulov2020** [comment]
Marks model index. Export model by right click. Print to log.

@wootguy you can test all now ?

**Comment by 2020karaulov2020** [comment]
First problem - I can't import bsp model to map (after merging player fall into ground at start round)

File: src/editor/Gui.cpp [hunk 3]
```diff
@@ -410,6 +417,30 @@ void Gui::draw3dContextMenus() {
 					command->execute();
 					app->pushUndoCommand(command);
 				}
+				if (ImGui::MenuItem("Export BSP model", 0, false, !app->isLoading)) {
+					if (app->pickInfo.modelIdx)
+					{
+						Bsp* tmpMap = new Bsp(map->path);
+						tmpMap->modelCount = 1;
+						tmpMap->models[0] = map->models[app->pickInfo.modelIdx];
+						STRUCTCOUNT removed = tmpMap->remove_unused_model_structures();
```
Comment (wootguy) [inline src/editor/Gui.cpp:426]: This should be called after deleting all the entities, so that the other BSP models are deleted.

File: src/editor/Gui.cpp [hunk 4]
```diff
@@ -410,6 +417,30 @@ void Gui::draw3dContextMenus() {
 					command->execute();
 					app->pushUndoCommand(command);
 				}
+				if (ImGui::MenuItem("Export BSP model", 0, false, !app->isLoading)) {
+					if (app->pickInfo.modelIdx)
+					{
+						Bsp* tmpMap = new Bsp(map->path);
+						tmpMap->modelCount = 1;
```
Comment (wootguy) [inline src/editor/Gui.cpp:424]: This doesn't need to be set if calling `removed_unused_model_structures` after deleting all entities.

File: src/editor/Gui.cpp [hunk 5]
```diff
@@ -586,22 +623,81 @@ void Gui::drawMenuBar() {
...
+										Entity* tmpEnt = tmpMap->ents[0];
+										for (int i = 1; i < tmpMap->ents.size(); i++)
+											delete tmpMap->ents[i];
+										tmpMap->ents.clear();
+										tmpMap->ents.push_back(tmpEnt);
+										tmpMap->delete_unused_hulls();
+										tmpMap->update_ent_lump();
+										tmpMap->update_lump_pointers();
+										logf("Export model %d to %s\n", i, (g_settings.gamedir + g_settings.workingdir + "model" + std::to_string(i) + ".bsp").c_str());
+										tmpMap->write(g_settings.gamedir + g_settings.workingdir + "model" + std::to_string(i) + ".bsp");
```
Comment (wootguy) [inline src/editor/Gui.cpp:681]: same comments here as above. Also this seems to mostly be duplicate code, so should move into its own method.

File: src/editor/Gui.cpp [hunk 6]
```diff
@@ -586,22 +623,81 @@ void Gui::drawMenuBar() {
...
+				for (int r = 0; r < app->mapRenderers.size(); r++)
+				{
+					if (!app->mapRenderers[r]->map)
+						continue;
+					if (ImGui::BeginMenu((std::string("Map ") + app->mapRenderers[r]->map->name + ".bsp").c_str())) {
+						Bsp* map = app->mapRenderers[r]->map;
+						if (map)
 						{
-							map->ents[0]->keyvalues["wad"] += map->name + ".wad" + ";";
+							for (int i = 0; i < map->modelCount; i++)
```
Comment (wootguy) [inline src/editor/Gui.cpp:659]: Can you add this to the context menu for entities instead? It would be a lot easier to select a model to export in the 3D view than to lookup model numbers.

image
Would fit best below `Transform` I guess.

File: src/editor/Gui.cpp [hunk 7]
```diff
@@ -2617,6 +2716,121 @@ void Gui::drawAbout() {
 	ImGui::End();
 }
 
+void Gui::drawMergeWindow() {
```
Comment (wootguy) [inline src/editor/Gui.cpp:2719]: can you leave the merging stuff out of this PR? There's a lot more work that needs to be done to get this working properly.

File: src/editor/Gui.h [hunk 8]
```diff
@@ -76,6 +78,7 @@ class Gui {
 	int fontSize = 22;
 	bool shouldReloadFonts = false;
 	bool shouldReloadTextureInfo = false;
+	bool settingLoaded = false;
```
Comment (wootguy) [inline src/editor/Gui.h:81]: `settingsLoaded` to be consistent

File: src/editor/Renderer.cpp [hunk 9]
```diff
@@ -1557,10 +1583,14 @@ void Renderer::setupView() {
...
 
-void Renderer::addMap(Bsp* map) {
+void Renderer::addMap(Bsp* map, bool real) {
 	BspRenderer* mapRenderer = new BspRenderer(map, bspShader, fullBrightBspShader, colorShader, pointEntRenderer);
 
 	mapRenderers.push_back(mapRenderer);
+	if (real)
+	{
+		foundRealMap = true;
+	}
```
Comment (wootguy) [inline src/editor/Renderer.cpp:1593]: why is this needed? Why aren't all maps "real" maps?

File: src/main.cpp [hunk 10]
```diff
@@ -632,6 +633,23 @@ void print_help(string command) {
...
+	}
+	else if (command == "editor" || command == "empty") {
+	logf(
+		"editor -\n"
+		"empty - Open empty bspguy window.\n\n"
+
+		"Usage:   bspguy editor\n"
+		"Usage:   bspguy empty\n"
+	);
+	}
```
Comment (wootguy) [inline src/main.cpp:652]: The tool isn't meant for creating new maps. I don't think this is needed for merging either.

File: src/util/util.cpp (hunk 2 above)
Comment (wootguy) [reply src/util/util.cpp:697]: Why TGA instead of BMP? All the asset-creation tools use BMP (Wally, sprite creation, model compiler).

File: src/editor/Gui.cpp [hunk 11]
```diff
@@ -586,22 +623,81 @@ void Gui::drawMenuBar() {
...
+				}
+			}
+
+			if (ImGui::MenuItem("Wavefront (.obj)", NULL)) {
+				Bsp* map = app->getMapContainingCamera()->map;
+				if (map)
+				{
+					map->ExportToObjWIP(g_settings.gamedir + g_settings.workingdir);
+				}
+			}
```
Comment (wootguy) [inline src/editor/Gui.cpp:648]: Rather than adding WIP to the method names, can you add WIP to the menu item label? The point is to make it clear to users that the feature isn't finished and might not work as expected.

Also would be nice to show a tooltip explaining what is/isn't working with this feature. Is it just textures and UV coordinates that are missing?

**Review by wootguy — CHANGES_REQUESTED** [review]
There is a lot going on in this PR. It would be better to make a PR for each new feature (export OBJ PR + export MDL PR + map import/merge PR). There's more that needs be done for map merging to work right, so I think that should be left out for now.

File: src/main.cpp (hunk 10 above)
Comment (2020karaulov2020) [reply src/main.cpp:652]: For example I don't like to open map from console, I not a linux user)))

File: src/editor/Renderer.cpp (hunk 9 above)
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:1593]: This needs for open map from import menu instead of console

File: src/editor/Gui.cpp (hunk 11 above)
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:648]: Yes need add export materials and write coordinates to obj file, i can try do it later. Now it just export map geometry

**Comment by 2020karaulov2020** [comment]
> feel free to ignore all those comments and create new, smaller PRs. I have a feeling this might turn into that other PR that got 100+ comments 😨

Okay, but after fix all problems from this pull request?

**Comment by 2020karaulov2020** [comment]
> feel free to ignore all those comments and create new, smaller PRs. I have a feeling this might turn into that other PR that got 100+ comments 😨

Okay, but after fix all problems from this pull request?

File: src/bsp/Bsp.cpp [hunk 12]
```diff
@@ -39,9 +41,9 @@ Bsp::Bsp() {
 			header.lump[i].nLength = 4096;
 			memset(lumps[i], 255, header.lump[i].nLength);
 		}
-		else 
+		else
 		{
-			lumps[i] = new byte[0];
+			lumps[i] = new byte[4]; // fix crash at replace_lump delete[]
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:46]: It strange but I got crash when do delete null data))

[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:46]: It strange but I got crash when do delete null data))

File: src/bsp/Bsp.cpp [hunk 13]
```diff
@@ -51,30 +53,64 @@ Bsp::Bsp() {
...
-	if (fpath.size() < 4 || fpath.rfind(".bsp") != fpath.size() - 4) {
-		fpath = fpath + ".bsp";
+	if (fpath.size() == 0)
+	{
+		this->init_empty_bsp();
 	}
-	this->path = fpath;
-	this->name = stripExt(basename(fpath));
-	valid = false;
+	else
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:66]: Ok

File: src/bsp/BspMerger.cpp [hunk 14]
```diff
@@ -825,8 +828,8 @@ bool BspMerger::merge(Bsp& mapA, Bsp& mapB) {
 
 	BSPPLANE separationPlane = separate(mapA, mapB);
 	if (separationPlane.nType == -1) {
-		logf("No separating axis found. The maps overlap and can't be merged.\n");
-		return false;
+		logf("Warning! No separating axis found. The maps be overlap.\n");
+		//return false;
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/bsp/BspMerger.cpp:832]: But i want to merge map with another's models, I can do it?(or how to create empty worldspawn to avoid this)

File: src/bsp/Bsp.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/bsp/Bsp.cpp:46]: hmm, maybe the lump pointer is updated wrong later.

File: src/bsp/BspMerger.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/bsp/BspMerger.cpp:832]: If you want the models to overlap, then the other map's models need to be submodels and used with entities instead of being merged into the worldspawn model. Otherwise you'd need to re-implement all of the compile tools to get proper BSP generation working.

File: src/editor/Renderer.cpp (hunk 9 above)
Comment (wootguy) [reply src/editor/Renderer.cpp:1593]: what makes the map "real"? Are empty maps not real?

File: src/main.cpp (hunk 10 above)
Comment (wootguy) [reply src/main.cpp:652]: I have it set up as my default program, so double-clicking a map opens the editor. You can also drag and drop a map onto the executable to open the editor.

Comment (wootguy) [reply src/main.cpp:652]: The problem is this program is both a command-line tool and a windows app. If no arguments are given, should you see the 3D view or a help menu? Depends on what you're trying to do. I guess the the command-line tools eventually shouldn't be needed, so I'm ok with the default mode opening an empty editor.

You added the "Import map" option so that allowing an empty editor makes sense, but I think that should be converted to an "Open map" option, since importing/merging will be confusing while it's unfinished. For instance, what happens when I copy from one map and paste in the other? Which map is saved when I click save? There's a lot to it and it deserves its own PR I think.

Comment (2020karaulov2020) [reply src/main.cpp:652]: > The problem is this program is both a command-line tool and a windows app. If no arguments are given, should you see the 3D view or a help menu? Depends on what you're trying to do. I guess the the command-line tools eventually shouldn't be needed, so I'm ok with the default mode opening an empty editor.
> 
> You added the "Import map" option so that allowing an empty editor makes sense, but I think that should be converted to an "Open map" option, since importing/merging will be confusing while it's unfinished. For instance, what happens when I copy from one map and paste in the other? Which map is saved when I click save? There's a lot to it and it deserves its own PR I think.

Yes, first imported map = open. I don't know how to open bsp guy program without any map, just run render without add map ?

File: src/editor/Renderer.cpp (hunk 9 above)
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:1593]: I can remove it and run render without any maps ?

File: src/bsp/BspMerger.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/bsp/BspMerger.cpp:836]: @wootguy  I tried it in new commit. But Unsuccessfully .... :(

**Comment by 2020karaulov2020** [comment]
@wootguy I tried to import exported model and merge it back! Unsuccessfully .... :(
https://github.com/wootguy/bspguy/pull/40/commits/a89f5abe138ac2fb872926f68bb859f2d6ad58be

**Comment by 2020karaulov2020** [comment]
@wootguy I tried to export model from cs_assault and import back. Exported model is valid. After import model placed at old position. But after merge maps, many models from cs_assault got random position and map is broken(players drop to ground). I think it be merged without problems, if exported without problems...but how I can fix it?((

**Comment by wootguy** [comment]
I think the problem is the model is being exported as the world model, which the merger will try to combine with the other map's world model. Origins are also ignored for the world model. If you just want to import it as an entity model then it should be exported at model index 1 or higher. I guess you'd need to create an empty-ish world model for the exported map too.

File: src/main.cpp (hunk 10 above)
Comment (wootguy) [reply src/main.cpp:652]: It should work ok without any map loaded, or at least be simpler to add logic for.

I was thinking the "Import" window could just be renamed to "Open", and it would replace the current map instead of adding a new one. At least, I think that would be better until map merging in the editor is finished.

File: src/editor/Renderer.cpp (hunk 9 above)
Comment (wootguy) [reply src/editor/Renderer.cpp:1593]: should be able to, yeah

**Comment by 2020karaulov2020** [comment]
> I think the problem is the model is being exported as the world model, which the merger will try to combine with the other map's world model. Origins are also ignored for the world model. If you just want to import it as an entity model then it should be exported at model index 1 or higher. I guess you'd need to create an empty-ish world model for the exported map too.

I already added option "merge as model" to add empty model, and save origin offset for  models,  you can check why it not works?
Exported models works in game (from entity, example func_wall with model "path_to_exportedmodel.bsp", but need precache model using cycler_sprite)

**Comment by 2020karaulov2020** [comment]
@wootguy after merging cs_assault with antenna model(exported and removed) , i got func_illusionar (door) at center of map)))

**Comment by 2020karaulov2020** [comment]
@wootguy  now model merged , but another models got unknown position, and if I see imported model - I see graphic problems )))

**Comment by 2020karaulov2020** [comment]
I see problem in SeperationPlane may be you can help me how to merge exported models back(I try to add empty model but it not helped) strange but i think problem in SeperationPlane, how i can skip it?

**Comment by wootguy** [comment]
Maps need a separating plane between worldspawn models or else they can't be merged. You'd probably need to offset model 0 so that it could be merged first, depending on how you're merging. Or maybe skip the BSP merging for model 0, and only merge the entity models from the exported map. There's probably a good amount of refactoring needed to support that properly.

**Comment by 2020karaulov2020** [comment]
> Maps need a separating plane between worldspawn models or else they can't be merged. You'd probably need to offset model 0 so that it could be merged first, depending on how you're merging. Or maybe skip the BSP merging for model 0, and only merge the entity models from the exported map. There's probably a good amount of refactoring needed to support that properly.

I can use "lifehack" import bsp models, but it needs two more entity(cycler_sprite for precache bsp model, and func_wall for showing bsp model in game with true geometry).
But better to merge it to one bsp, but I can't understand how your code works)) It really hard?)

**Comment by 2020karaulov2020** [comment]
@wootguy it really easy to export single bsp model, but why import cause problems... :((

**Comment by wootguy** [comment]
yeah sorry the code is kind of a mess. I don't know why the import is causing problems, and don't want to spend much time on the project right now either.

Another thing to try would be compiling a simple map with only 1 entity model, and comparing that to what you're exporting. If merging 2 full maps works, then I think there's something special about the export that's breaking the merge.

**Comment by 2020karaulov2020** [comment]
@wootguy I close this pull request and open clean !

File: src/editor/Renderer.cpp [hunk 15]
```diff
@@ -397,7 +397,13 @@ void Renderer::renderLoop() {
 		if (glfwGetTime( ) - lastTitleTime > 0.1)
 		{
 			lastTitleTime = glfwGetTime( );
-			glfwSetWindowTitle(window, std::string(std::string("bspguy - ") + getMapContainingCamera()->map->path).c_str());
+			if (BspRenderer * tmpRend = getMapContainingCamera())
+			{
+				if (tmpRend->map && tmpRend->map->name.size())
+				{
+					glfwSetWindowTitle(window, std::string(std::string("bspguy - ") + tmpRend->map->name + ".bsp").c_str());
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp]: I tried but... path to directory is more informative)) ??

File: src/editor/Gui.cpp [hunk 16]
```diff
@@ -599,6 +599,15 @@ void Gui::drawMenuBar() {
 					}
 				}
 			}
+
+			if (ImGui::MenuItem("Wavefront (.obj)", NULL)) {
+				Bsp* map = app->getMapContainingCamera()->map;
+				if (map)
+				{
+					map->ExportToObjWIP(g_settings.gamedir + g_settings.workingdir + map->name + ".obj");
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: Yes

File: src/editor/Gui.cpp (hunk 7 above)
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:2747]: I revert changes and it can be works as in command line)
